**Importing Dependencies**

In [1]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib

**Loading up the datasets and doing pre processing on it.**

In [2]:
df = pd.read_csv('/content/train.csv')


In [3]:
df.sample(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
1317  1318         120       FV         47.0     4230   Pave  Pave      Reg   
538    539          20       RL          NaN    11553   Pave   NaN      IR1   
90      91          20       RL         60.0     7200   Pave   NaN      Reg   
1209  1210          20       RL         85.0    10182   Pave   NaN      IR1   
686    687          60       FV         84.0    10207   Pave   NaN      Reg   
562    563          30       RL         63.0    13907   Pave   NaN      Reg   
436    437          50       RM         40.0     4400   Pave   NaN      Reg   
460    461          60       FV         75.0     8004   Pave   NaN      IR1   
1410  1411          60       RL         79.0    12420   Pave   NaN      Reg   
861    862         190       RL         75.0    11625   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal  \
1317         Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
538          Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
90           Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
1209         Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
686          Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
562          Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
436          Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
460          Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
1410         Lvl    AllPub  ...        0    NaN   NaN         NaN       0   
861          Lvl    AllPub  ...        0    NaN   NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice  
1317      4   2007       New        Partial     208900  
538       7   2006        WD         Normal     158000  
90        7   2006        WD         Normal     109900  
1209      5   2006       New        Partial     290000  
686       8   2007       New        Partial     227875  
562       7   2008        WD         Normal     108000  
436      10   2006        WD         Normal     116000  
460      12   2009       New        Partial     263435  
1410      6   2009        WD         Normal     230000  
861       4   2010        WD         Normal     131500  

[10 rows x 81 columns]

As the Column "id" is not relevant to for the training i am gonna drop that column

In [4]:
df.drop('Id', axis=1, inplace=True)


In [5]:
input_col = list(df.columns)[1:-1]
target_col = 'SalePrice'

In [6]:
x=df[input_col]
y=df[target_col]

Splitting the data into categorical and numerical


In [7]:
numeric_cols = x.select_dtypes(include=np.number).columns.tolist()
categorical_cols = x.select_dtypes('object').columns.tolist()

Handling the missing values

In [8]:
x[numeric_cols].isna().sum().sort_values(ascending=False).head(10)

LotFrontage    259
GarageYrBlt     81
MasVnrArea       8
OverallQual      0
LotArea          0
YearBuilt        0
OverallCond      0
BsmtFinSF1       0
BsmtFinSF2       0
BsmtUnfSF        0
dtype: int64

In [9]:
from sklearn.impute import SimpleImputer

In [10]:
Impute = SimpleImputer(strategy='mean')

In [11]:
Impute.fit(x[numeric_cols])

SimpleImputer()

In [12]:
Impute.transform(x[numeric_cols])

array([[6.500e+01, 8.450e+03, 7.000e+00, ..., 0.000e+00, 2.000e+00,
        2.008e+03],
       [8.000e+01, 9.600e+03, 6.000e+00, ..., 0.000e+00, 5.000e+00,
        2.007e+03],
       [6.800e+01, 1.125e+04, 7.000e+00, ..., 0.000e+00, 9.000e+00,
        2.008e+03],
       ...,
       [6.600e+01, 9.042e+03, 7.000e+00, ..., 2.500e+03, 5.000e+00,
        2.010e+03],
       [6.800e+01, 9.717e+03, 5.000e+00, ..., 0.000e+00, 4.000e+00,
        2.010e+03],
       [7.500e+01, 9.937e+03, 5.000e+00, ..., 0.000e+00, 6.000e+00,
        2.008e+03]])

In [13]:
x[numeric_cols].isna().sum().sort_values(ascending=False).head(10)

LotFrontage    259
GarageYrBlt     81
MasVnrArea       8
OverallQual      0
LotArea          0
YearBuilt        0
OverallCond      0
BsmtFinSF1       0
BsmtFinSF2       0
BsmtUnfSF        0
dtype: int64

In [14]:
from sklearn.preprocessing import OneHotEncoder

In [15]:
Encoder=OneHotEncoder( handle_unknown='ignore')
Encoder.fit(x[categorical_cols])

OneHotEncoder(handle_unknown='ignore')

In [16]:
# Fit and transform
encoded_data = Encoder.transform(x[categorical_cols])

# Convert to DataFrame
encoded_df = pd.DataFrame(encoded_data.toarray(), columns=Encoder.get_feature_names_out(categorical_cols))

# Drop original categorical columns
x = x.drop(columns=categorical_cols)

# Concatenate the new one-hot encoded columns
x = pd.concat([x, encoded_df], axis=1)

Scaling the numerical data

In [17]:
from sklearn.preprocessing import StandardScaler

In [18]:
Scaler=StandardScaler()
Scaler.fit(x[numeric_cols])

StandardScaler()

In [19]:
x[numeric_cols]=Scaler.transform(x[numeric_cols])

In [20]:
x

LotFrontage   LotArea  OverallQual  OverallCond  YearBuilt  \
0       -0.208034 -0.207142     0.651479    -0.517200   1.050994   
1        0.409895 -0.091886    -0.071836     2.179628   0.156734   
2       -0.084449  0.073480     0.651479    -0.517200   0.984752   
3       -0.414011 -0.096897     0.651479    -0.517200  -1.863632   
4        0.574676  0.375148     1.374795    -0.517200   0.951632   
...           ...       ...          ...          ...        ...   
1455    -0.331620 -0.260560    -0.071836    -0.517200   0.918511   
1456     0.615871  0.266407    -0.071836     0.381743   0.222975   
1457    -0.166839 -0.147810     0.651479     3.078570  -1.002492   
1458    -0.084449 -0.080160    -0.795151     0.381743  -0.704406   
1459     0.203918 -0.058112    -0.795151     0.381743  -0.207594   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  ...  \
0         0.878668    0.510015    0.575425   -0.288653  -0.944591  ...   
1        -0.429577   -0.572835    1.171992   -0.288653  -0.641228  ...   
2         0.830215    0.322174    0.092907   -0.288653  -0.301643  ...   
3        -0.720298   -0.572835   -0.499274   -0.288653  -0.061670  ...   
4         0.733308    1.360826    0.463568   -0.288653  -0.174865  ...   
...            ...         ...         ...         ...        ...  ...   
1455      0.733308   -0.572835   -0.973018   -0.288653   0.873321  ...   
1456      0.151865    0.084610    0.759659    0.722112   0.049262  ...   
1457      1.024029   -0.572835   -0.369871   -0.288653   0.701265  ...   
1458      0.539493   -0.572835   -0.865548    6.092188  -1.284176  ...   
1459     -0.962566   -0.572835    0.847389    1.509640  -0.976285  ...   

      SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  \
0                0.0           0.0           0.0          1.0   
1                0.0           0.0           0.0          1.0   
2                0.0           0.0           0.0          1.0   
3                0.0           0.0           0.0          1.0   
4                0.0           0.0           0.0          1.0   
...              ...           ...           ...          ...   
1455             0.0           0.0           0.0          1.0   
1456             0.0           0.0           0.0          1.0   
1457             0.0           0.0           0.0          1.0   
1458             0.0           0.0           0.0          1.0   
1459             0.0           0.0           0.0          1.0   

      SaleCondition_Abnorml  SaleCondition_AdjLand  SaleCondition_Alloca  \
0                       0.0                    0.0                   0.0   
1                       0.0                    0.0                   0.0   
2                       0.0                    0.0                   0.0   
3                       1.0                    0.0                   0.0   
4                       0.0                    0.0                   0.0   
...                     ...                    ...                   ...   
1455                    0.0                    0.0                   0.0   
1456                    0.0                    0.0                   0.0   
1457                    0.0                    0.0                   0.0   
1458                    0.0                    0.0                   0.0   
1459                    0.0                    0.0                   0.0   

      SaleCondition_Family  SaleCondition_Normal  SaleCondition_Partial  
0                      0.0                   1.0                    0.0  
1                      0.0                   1.0                    0.0  
2                      0.0                   1.0                    0.0  
3                      0.0                   0.0                    0.0  
4                      0.0                   1.0                    0.0  
...                    ...                   ...                    ...  
1455                   0.0                   1.0                    0.0  
1456          

In [21]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(x,y,test_size=0.2,random_state=42)

Lets Create The Model (Decision Tree)

In [22]:
from sklearn.tree import DecisionTreeRegressor

In [23]:
model=DecisionTreeRegressor()
model.fit(X_train, Y_train)

DecisionTreeRegressor()

lets evaluate the model


In [24]:
pred=model.predict(X_train)
pred2=model.predict(X_test)

In [25]:
print(Y_train.shape)
print(X_train.shape)

(1168,)
(1168, 302)


In [26]:
from sklearn.metrics import mean_squared_error
mse1 = mean_squared_error(Y_train, pred)
mse2 = mean_squared_error(Y_test, pred2)
rsme1=np.sqrt(mse1)
rsme2=np.sqrt(mse2)
print(f"Train Data Root Mean Squared Error: {rsme1}")
print(f"Test Data Root Mean Squared Error: {rsme2}")


Train Data Root Mean Squared Error: 372.996933403972
Test Data Root Mean Squared Error: 41949.18567057837


As you can see that the loss on train data is 0
while the loss on test data is arounf 4k,
this means the model has overfitted.
Lets try to fix this overfitting by hyperparameter tunning

In [27]:
def loss_calculate (pred,pred2):
  mse1 = mean_squared_error(Y_train, pred)
  mse2 = mean_squared_error(Y_test, pred2)
  rsme1=np.sqrt(mse1)
  rsme2=np.sqrt(mse2)
  print(f"Train Data Root Mean Squared Error: {rsme1}")
  print(f"Test Data Root Mean Squared Error: {rsme2}")

lets create a function to test out different parameters and their different values to minimize the loss.

In [28]:
def test_params(**params):
    model = DecisionTreeRegressor(random_state=42 , **params).fit(X_train, Y_train)
    pred = model.predict(X_train)
    pred2 = model.predict(X_test)
    loss_calculate(pred,pred2)


In [29]:
test_params(max_depth=10)

Train Data Root Mean Squared Error: 8750.612451832072
Test Data Root Mean Squared Error: 38632.98836701018


In [30]:
test_params(max_depth=15)

Train Data Root Mean Squared Error: 1607.9842809085285
Test Data Root Mean Squared Error: 43078.15484103531


In [31]:
test_params(max_depth=5)

Train Data Root Mean Squared Error: 28653.40336506412
Test Data Root Mean Squared Error: 39390.77804039401


In [32]:
test_params(max_depth=6)

Train Data Root Mean Squared Error: 23013.60520655602
Test Data Root Mean Squared Error: 38308.777047026124


So we get the minimum loss with max_depth = 6

In [33]:
test_params(max_depth=6,max_leaf_nodes=2**20)

Train Data Root Mean Squared Error: 23013.60520655602
Test Data Root Mean Squared Error: 35971.485358580714


So this is the minimum loss that we got to with hyperparameter tunning.
**Now lets try Random Forest Instead of Decision Tree**

In [36]:
from sklearn.ensemble import RandomForestRegressor
model=RandomForestRegressor()
model.fit(X_train, Y_train)

RandomForestRegressor()

In [38]:
model = RandomForestRegressor(random_state=42 ).fit(X_train, Y_train)
pred = model.predict(X_train)
pred2 = model.predict(X_test)
loss_calculate(pred,pred2)

Train Data Root Mean Squared Error: 11131.593182857936
Test Data Root Mean Squared Error: 29343.54552794146


thats a really good start, now lets tune hyperparameters and lets see how much we can minimize the loss

In [37]:
def test_params2(**params):
    model = RandomForestRegressor(random_state=42 , **params).fit(X_train, Y_train)
    pred = model.predict(X_train)
    pred2 = model.predict(X_test)
    loss_calculate(pred,pred2)

In [58]:
test_params2(n_estimators=250,max_depth=20,max_leaf_nodes=2**10)

Train Data Root Mean Squared Error: 11174.725593032359
Test Data Root Mean Squared Error: 28761.21020036789


**So we got to 28761 Root mean Squared Error with some hyperparameter tunning.**

In [61]:
y.sort_values(ascending=False).head(1)

691    755000
Name: SalePrice, dtype: int64

In [65]:
loss_perc= ( 28761.21020036789/   755000) * 100

In [69]:
print(f"The minimum RSME percentage of loss we got is: {loss_perc:.2f}%")

The minimum RSME percentage of loss we got is: 3.81%
